lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -428,93 +433,151 @@ static SILValue emitCodeForSymbolicValue(SymbolicValue symVal,
+      builder.createReleaseValue(loc, val, builder.getDefaultAtomicity());
+    }
+  };
+
+  // In the non-trival case, the foldedVal must at +1 as it must be a string
+  // constant. The originalVal could be at +0 or +1.
+
+  // Instructions that must be within the lifetime of the folded value. These
+  // are all the uses of the folded values that are introduced by folding.
+  SmallPtrSet<SILInstruction *, 4> lifetimeRangeOfFoldedVal;
+
+  // Iterate over all non-arc uses of originalVal and replace it with foldedVal,
+  // after emitting required ARC operation. The foldedVal should be at +1 even
+  // after replacing it in all the uses of originalVal.
+  while (Operand *use = getFirstNonARCUse(originalVal)) {
+    SILInstruction *user = use->getUser();
+
+    lifetimeRangeOfFoldedVal.insert(user);
+    // If this is an explicit borrow, include the end_borrows in the lifetime
+    // range of foldedVal.
+    if (BeginBorrowInst *beginBorrow = dyn_cast<BeginBorrowInst>(user)) {
+      for (EndBorrowInst *endBorrow : beginBorrow->getEndBorrows())
+        lifetimeRangeOfFoldedVal.insert(endBorrow);
+    }
 
-  if (!lifeTimeEndInstsOfOriginal.empty()) {
-    // Here, the original value is released, and so would be the folded value.
-    return;
-  }
+    // For a non-consuming use of originalVal, just replace the use by
+    // foldedVal and move on.
+    if (!isConsumingUse(use)) {
+      use->set(foldedVal);
+      continue;
+    }
 
-  // Here, the original value is not released. Release the folded value at the
-  // 'endOfUse' instructions passed as parameter.
-  bool hasOwnership = fun->hasOwnership();
-  for (SILInstruction *endInst : endOfUseInsts) {
-    SILBuilderWithScope builder(endInst);
-    if (hasOwnership) {
-      builder.createDestroyValue(endInst->getLoc(), foldedVal);
+    // For a consuming use of originalVal, destroy/release the originalVal,
+    // copy the foldedVal and use it in the place of originalVal.
+    SILBuilderWithScope builder(user);
+    SILLocation loc = user->getLoc();
+
+    releaseValue(builder, loc, originalVal);
+
+    SILValue copiedVal;
+    if (fun->hasOwnership()) {
+      copiedVal = builder.createCopyValue(loc, foldedVal);
     } else {
-      builder.createReleaseValue(endInst->getLoc(), foldedVal,
-                                 builder.getDefaultAtomicity());
+      builder.createRetainValue(loc, foldedVal, builder.getDefaultAtomicity());
+      copiedVal = foldedVal;
     }
+    use->set(copiedVal);
+  }
+
+  // Release the folded value at the end of the instructions in the lifetime
+  // range of folded value, which is given by the post-dominance frontier of
+  // lifetimeRangeOfFoldedVal.
+  SILInstruction *foldedValInst = foldedVal->getDefiningInstruction();
+  assert(foldedValInst);
+  SmallPtrSet<SILInstruction *, 1> defs = {foldedValInst};
+
+  llvm::SmallVector<SILInstruction *, 4> postDominators;
+  computePostDominatorsOfUsers(lifetimeRangeOfFoldedVal, defs, postDominators);
```

Comment:
This is just asking for the instructions at the lifetime boundary, which is exactly what ValueLifetimeAnalysis is supposed to do. That utility just needs to be slightly generalized to report the head of the "completion" blocks as part of the lifetime boundary. You don't need to do that in this PR, but as a heads up I'd like to change this call to `computePostDominatorsOfUsers` to `ValueLifetimeAnalysis`.